File: yeimu/lib/structure/sensor_reading.dart
class SensorReading {
  final List<int> accel;
  final List<int> gyro;

  SensorReading(this.accel, this.gyro);

  @override
  String toString() {
    return '$accel $gyro';
  }

  factory SensorReading.fromString(String string) {
    // unfortunately list to string conversion is printed with spaces between elements, so we instead split by list chars
    final List<String> accelAndGyroLists = string
        .split('[')
        .where((v) => v.isNotEmpty)
        .map((w) => w.replaceAll(RegExp(r'[,\]]'), ''))
        .toList();
    SensorReading reading = SensorReading(
      accelAndGyroLists[0] // accel list
          .split(' ') // split into individual axes
          .where((v) => v.isNotEmpty) // list of list string representations have two commas (??)
          .map((v) => int.parse(v)) // parse each value to int
          .toList(),
      //same again for gyro:
      accelAndGyroLists[1].split(' ').where((v) => v.isNotEmpty).map((v) => int.parse(v)).toList(),
    );

    if (reading.accel.length != 3 || reading.gyro.length != 3) {
      throw Exception('Invalid sensor reading: $reading');
    }

    return reading;
  }

  static String sensorReadingsToString(List<SensorReading> data) {
    return data.map((e) => e.toString()).join('\n');
  }

  static List<SensorReading> stringToSensorReadings(String data) {
    return data.split('\n').map((pair) => SensorReading.fromString(pair)).toList();
  }
}

File: yawnn/yawnnlib/structure/sensorReading.py
class SensorReading:
    """ A class representing a single reading from the sensor. """
    def __init__(self, accel : list[float], gyro : list[float]) -> None:
        self.accel = accel
        self.gyro = gyro
        
    @classmethod
    def fromString(cls, string : str) -> 'SensorReading':
        nonEmptySplits = filter(lambda v: v, string.split('['))
        accelAndGyroLists = list(map(lambda w: w.replace('[', '').replace(',', '').replace(']', ''), nonEmptySplits))
        
        m1 = map(float, filter(lambda v: v, accelAndGyroLists[0].split(' ')))
        m2 = map(float, filter(lambda v: v, accelAndGyroLists[1].split(' ')))
        
        reading = cls(list(m1), list(m2))
        assert len(reading.accel) == 3 and len(reading.gyro) == 3
        return reading
    
    def __eq__(self, __o: object) -> bool:
        if not isinstance(__o, SensorReading):
            return False
        return self.accel == __o.accel and self.gyro == __o.gyro